docs/design/proactive-diagnosis.md

```diff
@@ -0,0 +1,113 @@
+when encountering occasional exceptions. In the initial design, fault tolerance 
+during training was performed passively, meaning that only when a runtime 
+process or container failed and exited would the DLRover Master be notified and 
+attempt the corresponding fault-tolerant actions. However, as production scale 
+gradually increased, it became apparent that relying solely on passive fault 
+tolerance could not cover all exception scenarios. Therefore, the first 
+positive fault tolerance mechanism: 'the heartbeat mechanism' was implemented. 
+But considering that in distributed training, the training agent side also 
+employs a multi-process implementation, the aforementioned single heartbeat 
+mechanism cannot cover various exception scenarios that occur during actual 
+training operations (it can only cover some common cases).
+
+Therefore, a comprehensive, scalable framework is needed to gradually improve 
+the diagnosis and handling of corresponding exception scenarios.
+
+## Target
+
+- Be capable of proactive probing.
+- Requires flexible and scalable implementation.
+- Be able to reuse the current fault tolerance mechanisms.
+
+## Design
+
+The proactive diagnostic framework is mainly composed of the following roles:
+- [Diagnosis Manger](#Diagnosis-Manger)
+- [Diagnosis Agent](#Diagnosis-Agent)
+
+These components are distributed across the Master and Worker Nodes, 
+working together to achieve the main functionality of proactive diagnostics.
+
+<img src="../figures/diagnosis_arc.png" alt="Proactive Diagnostic Framework">
+
+Below is a more detailed introduction to the relevant concepts.
+
+### Diagnosis Manger
+The diagnosis-manager is implemented and runs in the Master node, with the main 
+responsibilities of driving pre-execution checks and runtime diagnostic logic(
+across all the workers).
+
+Runtime diagnostics primarily rely on an asynchronous scheduled thread and a 
+list of potential issues along with corresponding solutions(currently hardcoded). 
+During the training progress, this system continuously performs diagnostics and 
+determines the necessary actions to be executed. The specific diagnostic process 
+is shown in the following diagram, which is implemented through a simple inference chain. 
+The final actions can be either an event output or a specific fault-tolerance operation.
+
+<img src="../figures/diagnosis_manager_arc.png" alt="Diagnosis Manager">
+
+Then it will interact with other core roles of the master, such as the 
+job-manager, to assist in the lifecycle management of training.
+
+### Diagnosis Agent
+The diagnosis-agent is implemented and runs on the worker node, with the main 
+responsibilities of driving data collection on the training worker side, 
+executing some basic diagnostic logic, and reporting necessary information to 
+Master.
+
+<img src="../figures/diagnosis_agent_arc.png" alt="Diagnosis Agent">
+
+### Other Concepts
+#### Metric Collection
```

Comment:
Do you consider using Flight Recorder of torch >= 2.4.0 to diagnose hang?

Reply:
Not for now. Reason:

1. Need version >= 2.5.
2. Still a prototype feature.
3. The feature needs extra work if using Ascend device.

Reply:
Torch 2.4.0 works.